src/mattack_actors.cpp

```diff
@@ -153,82 +164,178 @@ void bite_actor::load( JsonObject &obj )
     max_mul = obj.get_float( "max_mul", 1.0f );
     move_cost = obj.get_int( "move_cost", 100 );
     accuracy = obj.get_int( "accuracy", INT_MIN );
-    no_infection_chance = obj.get_int( "no_infection_chance", 14 );
+
+    load_if_available( miss_msg_u, obj, "miss_msg_u", _( "The %s lunges at you, but you dodge!" ) );
+    load_if_available( miss_msg_npc, obj, "miss_msg_npc",
+                       _( "The %s lunges at <npcname>, but they dodge!" ) );
+
+    //~ 1$s is monster name, 2$s bodypart in accusative
+    load_if_available( hit_dmg_u, obj, "hit_dmg_u", _( "The %1$s bites your %2$s!" ) );
+    load_if_available( hit_dmg_npc, obj, "hit_dmg_npc", _( "The %1$s bites <npcname>'s %2$s!" ) );
+
+    //~ 1$s is monster name, 2$s bodypart in accusative
+    load_if_available( no_dmg_msg_u, obj, "no_dmg_msg_u",
+                       _( "The %1$s bites your %2$s, but fails to penetrate armor!" ) );
+    load_if_available( no_dmg_msg_npc, obj, "no_dmg_msg_npc",
+                       _( "The %1$s bites <npcname>'s %2$s, but fails to penetrate armor!" ) );
+
+    if( obj.has_array( "body_parts" ) ) {
+        JsonArray jarr = obj.get_array( "body_parts" );
+        while( jarr.has_more() ) {
+            JsonArray sub = jarr.next_array();
+            const body_part bp = get_body_part_token( sub.get_string( 0 ) );
+            const float prob = sub.get_float( 1 );
+            body_parts[ bp ] = prob;
+        }
+    }
+
+    if( obj.has_array( "effects" ) ) {
+        JsonArray jarr = obj.get_array( "effects" );
+        while( jarr.has_more() ) {
+            JsonObject eff = jarr.next_object();
+            effects.push_back( load_mon_effect_data( eff ) );
+        }
+    }
 }
 
-bool bite_actor::call( monster &z ) const
+Creature *melee_actor::find_target( monster &z ) const
 {
     if( !z.can_act() ) {
-        return false;
+        return nullptr;
     }
 
     Creature *target = z.attack_target();
     if( target == nullptr || !is_adjacent( z, *target ) ) {
+        return nullptr;
+    }
+
+    return target;
+}
+
+body_part roll_bp( const std::map<body_part, float> &c )
+{
+    if( c.empty() ) {
+        return num_bp;
+    }
+
+    const float sum = std::accumulate( c.begin(), c.end(), 0.0f,
+    []( float f, const decltype( *c.begin() ) &p ) {
+        return f + p.second;
+    } );
+
+    float roll = rng_float( 0.0f, sum );
```

Comment:
This looks suspiciously like the algorithm for a weighted list, shouldn't this use the `weighted_float_list` class?